Add time tracking for tasks

MillerJ20 opened the pull request:
This PR adds the following functionality:
- Pressing "s" on a task in the UI starts tracking the time
- Pressing "s" again stops tracking the task and calculates a total time spent tracking the task
- Completing a task that is being tracked will also stop tracking and calculate a total
- You can work on a task iteratively. You can stop tracking, then track again for a bit, the total calculation will track all of this and sum it together for a final time spent
- Multiple tasks can be tracked at once.

This is my first time coding in Lua (aside from configuring neovim), so let me know if there are any suggested changes. This was mostly written by hand, but I asked Zed to comment the code I added to make it more in line with the repo's conventions.

MillerJ20 committed:
Adding initial groundwork for tracking time. Next step is adding required values from the string, then extraction

MillerJ20 committed:
Task time tracking prototype

MillerJ20 committed:
Merge pull request #1 from MillerJ20/feature/time-tracking

Feature/time tracking

MillerJ20 committed:
time tracking fixes

MillerJ20 committed:
Numbers fix

MillerJ20 committed:
Finish implementation of time tracking logic

MillerJ20 committed:
Adding comments to code

MillerJ20 committed:
Remove unused code in due_helpers

MillerJ20 committed:
Formatting fix

MillerJ20 committed:
Formatting fix pt.2

MillerJ20 committed:
Formatting fix pt.3

MillerJ20 committed:
Formatting fix finished for real this time

MillerJ20 committed:
help

MillerJ20 committed:
Added unit tests, clean up start/end when done tracking, change total to tracked, write at end of line

cche committed:
Eliminated spaces in tracked time to adhere to todo.txt format.

Eliminated the end:1234567890 tag that is not used anymore.
Use local end time instead to calculate tracked time.
Adjusted tests to modified tracked time format.

cche committed:
Eliminated spaces in tracked time to adhere to todo.txt format.

Eliminated the end:1234567890 tag that is not used anymore.
Use local end time instead to calculate tracked time.
Adjusted tests to modified tracked time format.

cche committed:
Merge remote-tracking branch 'origin/pr-2-time-tracking' into pr-2-time-tracking

cche committed:
Merge branch 'main' into main

MillerJ20 committed:
:Merge remote-tracking branch 'upstream/pr-2-time-tracking'

MillerJ20 committed:
Updated tests, pulled in changes from pr-2 branch

MillerJ20 committed:
Fixed merge conflict

cche reviewed (changes requested):
Hi @MillerJ20 ,

With this last commit you introduced spaces after the h, m and s which should be eliminated.

Also there are tests that do not pass because extract_tracked_time returns "tracked:XhXmXs" and not hours, minutes, seconds, so in lines 323 and 332 you can do:
      local hours, minutes, seconds = parser.extract_tracked_time(line):match("tracked:(%d+)h(%d+)m(%d+)s")

With those changes, it is ready to merge.

Thanks! I can't wait to use this!

cche committed:
fix: Correct time extraction inconsistency

In lua/todo-txt/task.lua lua/todo-txt/util.lua tests/time_tracking_spec.lua the function
extract_tracked_time does not return hours, minutes and seconds, so we
have to explicitely extract them from the string.
